src/Libraries/Microsoft.Extensions.DataIngestion/Chunkers/SemanticSimilarityChunker.cs

```diff
@@ -0,0 +1,142 @@
+/// </summary>
+public sealed class SemanticSimilarityChunker : IngestionChunker<string>
+{
+    private readonly ElementsChunker _elementsChunker;
+    private readonly IEmbeddingGenerator<string, Embedding<float>> _embeddingGenerator;
+    private readonly float _thresholdPercentile;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SemanticSimilarityChunker"/> class.
+    /// </summary>
+    /// <param name="embeddingGenerator">Embedding generator.</param>
+    /// <param name="options">The options for the chunker.</param>
+    /// <param name="thresholdPercentile">Threshold percentile to consider the chunks to be sufficiently similar.</param>
+    public SemanticSimilarityChunker(
+        IEmbeddingGenerator<string, Embedding<float>> embeddingGenerator,
+        IngestionChunkerOptions options,
+        float thresholdPercentile = 95.0f)
+    {
+        _embeddingGenerator = Throw.IfNull(embeddingGenerator);
+        _elementsChunker = new(options);
+
+        if (thresholdPercentile < 0f || thresholdPercentile > 100f)
+        {
+            Throw.ArgumentOutOfRangeException(nameof(thresholdPercentile), "Threshold percentile must be between 0 and 100.");
+        }
+
+        _thresholdPercentile = thresholdPercentile;
+    }
+
+    /// <inheritdoc/>
+    public override async IAsyncEnumerable<IngestionChunk<string>> ProcessAsync(IngestionDocument document,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        _ = Throw.IfNull(document);
+
+        List<(IngestionDocumentElement, float)> distances = await CalculateDistancesAsync(document, cancellationToken).ConfigureAwait(false);
+        foreach (var chunk in MakeChunks(document, distances))
+        {
+            yield return chunk;
+        }
+    }
+
+    private async Task<List<(IngestionDocumentElement element, float distance)>> CalculateDistancesAsync(IngestionDocument documents, CancellationToken cancellationToken)
+    {
+        List<(IngestionDocumentElement element, float distance)> elementDistance = [];
+        List<string> semanticContents = [];
+
+        foreach (IngestionDocumentElement element in documents.EnumerateContent())
+        {
+            string? semanticContent = element is IngestionDocumentImage img
+                ? img.AlternativeText ?? img.Text
+                : element.GetMarkdown();
+
+            if (!string.IsNullOrEmpty(semanticContent))
+            {
+                elementDistance.Add((element, default));
+                semanticContents.Add(semanticContent!);
+            }
+        }
+
+        if (elementDistance.Count > 0)
+        {
+            var embeddings = await _embeddingGenerator.GenerateAsync(semanticContents, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            for (int i = 0; i < elementDistance.Count - 1; i++)
```

Comment:
What happens if elementDistance.Count == 1? Is `default` the right distance value for elementDistance[0].distance?